File: src/utils/notice-period-calculator.ts
/**
 * Malta Notice Period Calculator
 * Based on Employment and Industrial Relations Act - Article 36(5)
 */

export interface NoticePeriodInput {
    /** Total months of continuous service */
    monthsOfService: number;
    /** Whether employee is in probation period */
    isInProbation?: boolean;
}

export interface NoticePeriodOutput {
    /** Notice period in weeks */
    weeks: number;
    /** Notice period in calendar days (approximate) */
    days: number;
    /** Description of the service bracket */
    serviceBracket: string;
    /** Whether during probation */
    isInProbation: boolean;
}

/**
 * Notice period brackets as per Malta Employment and Industrial Relations Act
 * Article 36(5) - Indefinite contracts
 */
const NOTICE_PERIOD_BRACKETS = [
    { minMonths: 0, maxMonths: 1, weeks: 0, description: "Less than 1 month (Probation)" },
    { minMonths: 1, maxMonths: 6, weeks: 1, description: "1 to 6 months" },
    { minMonths: 6, maxMonths: 24, weeks: 2, description: "6 months to 2 years" },
    { minMonths: 24, maxMonths: 48, weeks: 4, description: "2 to 4 years" },
    { minMonths: 48, maxMonths: 84, weeks: 8, description: "4 to 7 years" },
    { minMonths: 84, maxMonths: 96, weeks: 9, description: "7 to 8 years" },
    { minMonths: 96, maxMonths: 108, weeks: 10, description: "8 to 9 years" },
    { minMonths: 108, maxMonths: 120, weeks: 11, description: "9 to 10 years" },
    { minMonths: 120, maxMonths: Infinity, weeks: 12, description: "More than 10 years" },
] as const;

/**
 * Calculate notice period based on months of service
 */
export function calculateNoticePeriod(input: NoticePeriodInput): NoticePeriodOutput {
    const { monthsOfService, isInProbation = false } = input;

    // During probation with less than 1 month: no notice required
    if (isInProbation && monthsOfService < 1) {
        return {
            weeks: 0,
            days: 0,
            serviceBracket: "Less than 1 month (Probation)",
            isInProbation: true,
        };
    }

    // During probation with more than 1 month: 1 week notice
    if (isInProbation && monthsOfService >= 1) {
        return {
            weeks: 1,
            days: 7,
            serviceBracket: "Probation period (more than 1 month)",
            isInProbation: true,
        };
    }

    // Find the appropriate bracket
    const bracket = NOTICE_PERIOD_BRACKETS.find(
        (b) => monthsOfService >= b.minMonths && monthsOfService < b.maxMonths
    );

    if (!bracket) {
        // Default to maximum (10+ years)
        return {
            weeks: 12,
            days: 84,
            serviceBracket: "More than 10 years",
            isInProbation: false,
        };
    }

    return {
        weeks: bracket.weeks,
        days: bracket.weeks * 7,
        serviceBracket: bracket.description,
        isInProbation: false,
    };
}

/**
 * Convert years and months to total months
 */
export function toMonths(years: number, months: number): number {
    return years * 12 + months;
}

/**
 * Format notice period for display
 */
export function formatNoticePeriod(weeks: number): string {
    if (weeks === 0) return "No notice required";
    if (weeks === 1) return "1 week";
    return `${weeks} weeks`;
}

File: ios-app/Packages/CalculationKit/Sources/CalculationKit/NoticePeriod/NoticePeriodCalculator.swift
import Foundation

/// Malta Notice Period Calculator.
///
/// Determines the minimum notice period for indefinite contracts based on
/// continuous service, per the Employment and Industrial Relations Act
/// Article 36(5). Special rules apply during probation.
public struct NoticePeriodCalculator: Sendable {

    /// Creates a new notice period calculator.
    public init() {}

    /// Calculates the required notice period.
    ///
    /// - Parameter input: A ``NoticePeriodInput`` with months of service and probation flag.
    /// - Returns: A ``NoticePeriodOutput`` with weeks, days, and bracket description.
    public func calculate(input: NoticePeriodInput) -> NoticePeriodOutput {
        // Probation: less than 1 month — no notice required
        if input.isInProbation, input.monthsOfService < 1 {
            return NoticePeriodOutput(
                weeks: 0,
                days: 0,
                serviceBracket: "Less than 1 month (Probation)",
                isInProbation: true
            )
        }

        // Probation: 1 month or more — 1 week notice
        if input.isInProbation, input.monthsOfService >= 1 {
            return NoticePeriodOutput(
                weeks: 1,
                days: 7,
                serviceBracket: "Probation period (more than 1 month)",
                isInProbation: true
            )
        }

        // Standard brackets (Article 36(5))
        let bracket = Self.findBracket(months: input.monthsOfService)
        return NoticePeriodOutput(
            weeks: bracket.weeks,
            days: bracket.weeks * 7,
            serviceBracket: bracket.description,
            isInProbation: false
        )
    }

    // MARK: - Private Helpers

    /// A notice-period bracket defined by a service-month range.
    private struct Bracket {
        /// Inclusive lower bound in months.
        let minMonths: Int
        /// Exclusive upper bound in months (use `Int.max` for open-ended).
        let maxMonths: Int
        /// Required notice in weeks.
        let weeks: Int
        /// Human-readable bracket label.
        let description: String
    }

    /// Ordered list of notice-period brackets per Malta law.
    private static let brackets: [Bracket] = [
        Bracket(minMonths: 0,   maxMonths: 1,       weeks: 0,  description: "Less than 1 month"),
        Bracket(minMonths: 1,   maxMonths: 6,       weeks: 1,  description: "1 to 6 months"),
        Bracket(minMonths: 6,   maxMonths: 24,      weeks: 2,  description: "6 months to 2 years"),
        Bracket(minMonths: 24,  maxMonths: 48,      weeks: 4,  description: "2 to 4 years"),
        Bracket(minMonths: 48,  maxMonths: 84,      weeks: 8,  description: "4 to 7 years"),
        Bracket(minMonths: 84,  maxMonths: 96,      weeks: 9,  description: "7 to 8 years"),
        Bracket(minMonths: 96,  maxMonths: 108,     weeks: 10, description: "8 to 9 years"),
        Bracket(minMonths: 108, maxMonths: 120,     weeks: 11, description: "9 to 10 years"),
        Bracket(minMonths: 120, maxMonths: Int.max,  weeks: 12, description: "More than 10 years"),
    ]

    /// Finds the matching bracket for a given number of service months.
    private static func findBracket(months: Int) -> Bracket {
        for bracket in brackets where months >= bracket.minMonths && months < bracket.maxMonths {
            return bracket
        }
        // Fallback to the maximum bracket
        return brackets[brackets.count - 1]
    }
}
